src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/CodeActions/RoslynCodeActionHelpers.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor;
+using Microsoft.CodeAnalysis.Razor.CodeActions;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using ExternalHandlers = Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost.Handlers;
+using RoslynTextEdit = Roslyn.LanguageServer.Protocol.TextEdit;
+
+namespace Microsoft.CodeAnalysis.Remote.Razor;
+
+[Export(typeof(IRoslynCodeActionHelpers)), Shared]
+internal sealed class RoslynCodeActionHelpers() : IRoslynCodeActionHelpers
+{
+    public async Task<string?> GetFormattedNewFileContentsAsync(IProjectSnapshot projectSnapshot, Uri csharpFileUri, string newFileContent, CancellationToken cancellationToken)
+    {
+        Debug.Assert(projectSnapshot is RemoteProjectSnapshot);
+        var project = ((RemoteProjectSnapshot)projectSnapshot).Project;
+
+        var document = project.AddDocument(RazorUri.GetDocumentFilePathFromUri(csharpFileUri), newFileContent);
+
+        return await ExternalHandlers.CodeActions.GetFormattedNewFileContentAsync(document, cancellationToken).ConfigureAwait(false);
+    }
```

Comment:
Consider not creating a new async state machine.

```suggestion
    public Task<string?> GetFormattedNewFileContentsAsync(IProjectSnapshot projectSnapshot, Uri csharpFileUri, string newFileContent, CancellationToken cancellationToken)
    {
        Debug.Assert(projectSnapshot is RemoteProjectSnapshot);
        var project = ((RemoteProjectSnapshot)projectSnapshot).Project;

        var document = project.AddDocument(RazorUri.GetDocumentFilePathFromUri(csharpFileUri), newFileContent);

        return ExternalHandlers.CodeActions.GetFormattedNewFileContentAsync(document, cancellationToken);
    }
```

Reply:
Okay, but I need to employ a `!` because the Roslyn method returns `Task<string>`, so its not pretty. Unless I'm forgetting some other way to avoid the nullability compiler warning?

Reply:
Ahhh... I see. Yeah, that's unfortunate. However, if the Roslyn method returns a `Task<string>`, why does this return a `Task<string?>`? Is the Roslyn method not annotated correctly? Or, can this also return a `Task<string>`?

Reply:
It's the non-cohost impl. It makes an LSP call, so can always be null.

Reply:
Had a shower thought, I can make this non-nullable by moving the bit that handles null into the non-cohost impl. Thank you for the brain spark :D